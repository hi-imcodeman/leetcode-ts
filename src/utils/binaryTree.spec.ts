import {
    insertLevelOrder,
    inOrder,
    preOrder,
    postOrder,
    levelOrder,
    getAsciiTree
} from './binaryTree'
import {flatten} from 'lodash'


describe('Binary Tree Utils', () => {
    test('insertLevelOrder', () => {
        const bt = insertLevelOrder([1, 2, 3, 4, 5, 6, 6, 6, 6])
        expect(bt).toMatchSnapshot()
    })
    test('InOrder Traversal', () => {
        const bt = insertLevelOrder([1, 2, 3, 4, 5, 6, 6, 6, 6])
        const inOrderArr = inOrder(bt)
        expect(inOrderArr).toMatchSnapshot()
    })
    test('PreOrder Traversal', () => {
        const bt = insertLevelOrder([1, 2, 3, 4, 5, 6, 6, 6, 6])
        const preOrderArr = preOrder(bt)
        expect(preOrderArr).toMatchSnapshot()
    })
    test('PostOrder Traversal', () => {
        const bt = insertLevelOrder([1, 2, 3, 4, 5, 6, 6, 6, 6])
        const postOrderArr = postOrder(bt)
        expect(postOrderArr).toMatchSnapshot()
    })
    test('LevelOrder Traversal', () => {
        const bt = insertLevelOrder([1, 2, 3, 4, 5, 6, 6, 6, 6])
        const levelOrderArr = levelOrder(bt)
        expect(levelOrderArr).toMatchSnapshot()
    })
    test('LevelOrder Insert and Traversal with null', () => {
        const bt = insertLevelOrder([1, null, 2, null, null, 3])
        expect(bt).toMatchSnapshot('BT')
        const levelOrderArr = levelOrder(bt)
        expect(levelOrderArr).toMatchSnapshot('levelOrderArr')
    })
    test('Traversals with string', () => {
        const bt = insertLevelOrder(['a', 'b', 'c', null, 'e', 'f'])
        expect(bt).toMatchSnapshot('BT')
        const inOrderArr = inOrder(bt)
        const preOrderArr = preOrder(bt)
        const postOrderArr = postOrder(bt)
        expect(inOrderArr).toMatchSnapshot('inOrder')
        expect(preOrderArr).toMatchSnapshot('preOrder')
        expect(postOrderArr).toMatchSnapshot('postOrder')
    })
    test('Get ASCII Tree', () => {
        const asciiTree= getAsciiTree([3,5,1,6,2,0,8,null,null,7,4])
        expect(asciiTree).toMatchSnapshot()
    })

    test('Construct ASCII tree from BT', () => {
        const treeVals = [3,5,1,6,2,0,8,null,null,7,4]
        const asciiTreeFromArr= getAsciiTree(treeVals)
        const bt= insertLevelOrder(treeVals)
        const arr=levelOrder(bt)
        // arr.pop()
        const asciiTreeFromBT=getAsciiTree(flatten(arr))
        expect(asciiTreeFromBT).toBe(asciiTreeFromArr)
    })
})