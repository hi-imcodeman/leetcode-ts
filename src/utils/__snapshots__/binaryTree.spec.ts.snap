// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Binary Tree Utils InOrder Traversal 1`] = `
Array [
  6,
  4,
  6,
  2,
  5,
  1,
  6,
  3,
  6,
]
`;

exports[`Binary Tree Utils LevelOrder Insert and Traversal with null: BT 1`] = `
TreeNode {
  "left": TreeNode {
    "left": TreeNode {
      "left": null,
      "right": null,
      "val": null,
    },
    "right": TreeNode {
      "left": null,
      "right": null,
      "val": null,
    },
    "val": null,
  },
  "right": TreeNode {
    "left": TreeNode {
      "left": null,
      "right": null,
      "val": 3,
    },
    "right": null,
    "val": 2,
  },
  "val": 1,
}
`;

exports[`Binary Tree Utils LevelOrder Insert and Traversal with null: levelOrderArr 1`] = `
Array [
  Array [
    1,
  ],
  Array [
    null,
    2,
  ],
  Array [
    null,
    null,
    3,
  ],
]
`;

exports[`Binary Tree Utils LevelOrder Traversal 1`] = `
Array [
  Array [
    1,
  ],
  Array [
    2,
    3,
  ],
  Array [
    4,
    5,
    6,
    6,
  ],
  Array [
    6,
    6,
  ],
]
`;

exports[`Binary Tree Utils PostOrder Traversal 1`] = `
Array [
  6,
  6,
  4,
  5,
  2,
  6,
  6,
  3,
  1,
]
`;

exports[`Binary Tree Utils PreOrder Traversal 1`] = `
Array [
  1,
  2,
  4,
  6,
  6,
  5,
  3,
  6,
  6,
]
`;

exports[`Binary Tree Utils Traversals with string: BT 1`] = `
TreeNode {
  "left": TreeNode {
    "left": TreeNode {
      "left": null,
      "right": null,
      "val": null,
    },
    "right": TreeNode {
      "left": null,
      "right": null,
      "val": "e",
    },
    "val": "b",
  },
  "right": TreeNode {
    "left": TreeNode {
      "left": null,
      "right": null,
      "val": "f",
    },
    "right": null,
    "val": "c",
  },
  "val": "a",
}
`;

exports[`Binary Tree Utils Traversals with string: inOrder 1`] = `
Array [
  null,
  "b",
  "e",
  "a",
  "f",
  "c",
]
`;

exports[`Binary Tree Utils Traversals with string: postOrder 1`] = `
Array [
  null,
  "e",
  "b",
  "f",
  "c",
  "a",
]
`;

exports[`Binary Tree Utils Traversals with string: preOrder 1`] = `
Array [
  "a",
  "b",
  null,
  "e",
  "c",
  "f",
]
`;

exports[`Binary Tree Utils insertLevelOrder 1`] = `
TreeNode {
  "left": TreeNode {
    "left": TreeNode {
      "left": TreeNode {
        "left": null,
        "right": null,
        "val": 6,
      },
      "right": TreeNode {
        "left": null,
        "right": null,
        "val": 6,
      },
      "val": 4,
    },
    "right": TreeNode {
      "left": null,
      "right": null,
      "val": 5,
    },
    "val": 2,
  },
  "right": TreeNode {
    "left": TreeNode {
      "left": null,
      "right": null,
      "val": 6,
    },
    "right": TreeNode {
      "left": null,
      "right": null,
      "val": 6,
    },
    "val": 3,
  },
  "val": 1,
}
`;
